### ISSUE TITLE
mysql specific sql

### ISSUE THREAD

**Issue by danmarsden** [NMAI-lab/moodle-mod_socialwiki#38]
When using aggregate functions (count, max etc) all the result columns that are not aggregates should appear in the group by clause.
This query will work in MySql (which does not follow the standard) but will fail in other supported databases like postgres, MS Sql
https://github.com/NMAI-lab/moodle-mod_socialwiki/blob/master/locallib.php#L417

here's a stackoverflow post that I've referenced before to help explain it:
http://stackoverflow.com/questions/4611897/group-by-aggregate-function-confusion-in-sql

**Comment by danmarsden** [comment]
here's another example from your code:
https://github.com/NMAI-lab/moodle-mod_socialwiki/blob/master/locallib.php#L450

you should look at all the queries you have written that include aggregates and adjust them as required. (even better would be for you to install PostGres and test your plugin using PostGres)

**Comment by bencroskery** [comment]
I haven't had the chance yet to set up a Moodle site with postgres but I've done a little bit of testing with some queries and I don't seem to be encountering any errors.
Maybe the site I'm using just isn't catching the error for whatever reason but both of the queries that you mentioned worked without any issues with some test tables I made up.

You should be able to see them through these links:
http://sqlfiddle.com/#!15/9c488/3
http://sqlfiddle.com/#!15/d3acf/1

From what I can tell from the stackoverflow you linked, the problem in SQL other than mysql is that * should not be used and instead all of the column should be listed (ex. p.id, p.subwikiid, p.title, etc.)

**Comment by danmarsden** [comment]
don't know anything about sqlfiddle but the SQL standards say that if you use an aggregate field in your select, all the columns in your select that don't form part of the aggregate should be part of the group by statement. MySQl ignores this but it causes issues with other db types that Moodle supports. it's not the * that is the issue it's that you have columns in your SELECT (caused by the *) that should be in the group by.

With larger plugins like yours it is really better for the developer to install/test using PostGres because if it works in PostGres it is likely to work in all dbs including MySql/MS Sql.

**Comment by bencroskery** [comment]
I've done testing with a server running PostGres and haven't run into any issues. The functions that you mentioned are exclusively for the search, which hasn't had any errors appear and seems to returns the correct pages in all the cases that I tested.

I did some SQL command testing and I think that problem occurs here if the command has SELECT *, but in this case p.* has no duplicates and won't run into any errors.
